src/sagemaker/workflow/steps.py

```diff
@@ -148,9 +159,51 @@ def _resolve_depends_on(
             elif isinstance(step, str):
                 depends_on.append(step)
             else:
-                raise ValueError(f"Invalid input step name: {step}")
+                raise ValueError(f"Invalid input step type: {type(step)}")
         return depends_on
 
+    def _find_step_dependencies(self, step_map):
+        """Find the step names this step is dependent on."""
+        # import here to prevent circular import
+        from sagemaker.workflow.step_collections import StepCollection
+
+        def _find_dependencies_in_step_arguments(obj):
+            dependencies = set()
+            if isinstance(obj, dict):
+                for value in obj.values():
+                    if isinstance(value, (PipelineVariable, Condition)):
+                        dependencies.update(map(_get_step_name_from_str, value._referenced_steps))
+                    dependencies.update(_find_dependencies_in_step_arguments(value))
+            elif isinstance(obj, list):
+                for item in obj:
+                    if isinstance(item, (PipelineVariable, Condition)):
+                        dependencies.update(map(_get_step_name_from_str, item._referenced_steps))
+                    dependencies.update(_find_dependencies_in_step_arguments(item))
+            return dependencies
+
+        def _find_dependencies_in_depends_on_list():
```

Comment:
nit: This func has grown too long , can possibly move out all these internal functions? Should also help in Unit testing?